akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ReplayingEvents.scala

```diff
@@ -131,93 +131,94 @@ private[akka] final class ReplayingEvents[C, E, S](
+                replication.clearContext()
+              case None =>
             }
 
-            eventSeq match {
-              case SingleEventSeq(event) => handleEvent(event)
-              case EventsSeq(events)     => events.foreach(handleEvent)
-              case EmptyEventSeq         => // no events
+            replicatedMetaAndSelfReplica match {
+              case Some((meta, selfReplica, replication)) if meta.originReplica != selfReplica =>
+                // keep track of highest origin seqnr per other replica
+                state = state.copy(
+                  state = newState,
+                  eventSeenInInterval = true,
+                  version = meta.version,
+                  seenSeqNrPerReplica = state.seenSeqNrPerReplica + (meta.originReplica -> meta.originSequenceNr))
+                replication.clearContext()
+              case Some((_, _, replication)) =>
+                replication.clearContext()
+                state = state.copy(state = newState, eventSeenInInterval = true)
+              case _ =>
+                state = state.copy(state = newState, eventSeenInInterval = true)
             }
+          }
 
-            this
-          } catch {
-            case NonFatal(ex) =>
-              state = state.copy(repr.sequenceNr)
-              onRecoveryFailure(ex, eventForErrorReporting.toOption)
+          eventSeq match {
+            case SingleEventSeq(event) => handleEvent(event)
+            case EventsSeq(events)     => events.foreach(handleEvent)
+            case EmptyEventSeq         => // no events
           }
 
-        case RecoverySuccess(highestJournalSeqNr) =>
+          this
+        } catch {
+          case NonFatal(ex) =>
+            state = state.copy(repr.sequenceNr)
+            onRecoveryFailure(ex, eventForErrorReporting.toOption, "replaying-event")
+        }
+
+      case RecoverySuccess(highestJournalSeqNr) =>
+        try {
           val highestSeqNr = Math.max(highestJournalSeqNr, state.seqNr)
           state = state.copy(seqNr = highestSeqNr)
           setup.internalLogger.debug("Recovery successful, recovered until sequenceNr: [{}]", highestSeqNr)
           onRecoveryCompleted(state)
+        } catch {
+          case ex: UnstashException[_] =>
+            // let supervisor handle it, don't treat as recovery failure
+            throw ex
 
-        case ReplayMessagesFailure(cause) =>
-          onRecoveryFailure(cause, Some(response))
+          case NonFatal(cause) =>
+            onRecoveryFailure(cause, None, "recovery-completed")
+        }
 
-        case _ =>
-          Behaviors.unhandled
-      }
-    } catch {
-      case ex: UnstashException[_] =>
-        // let supervisor handle it, don't treat it as recovery failure
-        throw ex
-      case NonFatal(cause) =>
-        onRecoveryFailure(cause, None)
+      case ReplayMessagesFailure(cause) =>
+        onRecoveryFailure(cause, Some(response), "failure-from-journal")
```

Comment:
If this threw previously, that would lead to a new call to onRecoveryFailure, but maybe it makes more sense that it doesn't.

Reply:
It shouldn't throw since we have:
```
setup.onSignal(state.state, RecoveryFailed(cause), catchAndLog = true)
```

but much other code there as well, which hopefully doesn't throw. Should we guard against that with an additional catch inside onRecoveryFailure?

It should always end with:
```
throw new JournalFailureException(msg, cause)
```